packages/react-reconciler/src/ReactFiberThrow.js

```diff
@@ -556,15 +558,55 @@ function throwException(
 
       // Even though the user may not be affected by this error, we should
       // still log it so it can be fixed.
-      queueHydrationError(createCapturedValueAtFiber(value, sourceFiber));
+      if (value !== HydrationMismatchException) {
+        const wrapperError = new Error(
+          'There was an error while hydrating but React was able to recover by ' +
+            'instead client rendering from the nearest Suspense boundary.',
+          {cause: value},
+        );
+        queueHydrationError(
+          createCapturedValueAtFiber(wrapperError, sourceFiber),
+        );
+      }
+      return false;
+    } else {
+      if (value !== HydrationMismatchException) {
+        const wrapperError = new Error(
+          'There was an error while hydrating but React was able to recover by ' +
+            'instead client rendering the entire root.',
+          {cause: value},
+        );
+        queueHydrationError(
+          createCapturedValueAtFiber(wrapperError, sourceFiber),
+        );
+      }
+      const workInProgress: Fiber = (root.current: any).alternate;
+      // Schedule an update at the root to log the error but this shouldn't
+      // actually happen because we should recover.
+      workInProgress.flags |= ShouldCapture;
+      const lane = pickArbitraryLane(rootRenderLanes);
+      workInProgress.lanes = mergeLanes(workInProgress.lanes, lane);
+      const rootErrorInfo = createCapturedValueAtFiber(value, sourceFiber);
+      const update = createRootErrorUpdate(
+        workInProgress.stateNode,
+        rootErrorInfo, // This should never actually get logged due to the recovery.
+        lane,
+      );
+      enqueueCapturedUpdate(workInProgress, update);
+      renderDidError();
       return false;
     }
   } else {
     // Otherwise, fall through to the error path.
   }
 
-  value = createCapturedValueAtFiber(value, sourceFiber);
-  renderDidError(value);
+  const wrapperError = new Error(
+    'There was an error during concurrent rendering but React was able to recover by ' +
+      'instead synchronously rendering the entire root.',
```

Comment:
Technically I think we could error in the sync pass and then recover in the second sync pass. We could check if this is actually a concurrent render first but we don't really expect this case.